### DISCUSSION TITLE
Help adding easy-hls-nix to devos

### DISCUSSION THREAD

**Discussion by codygman** [nrdxp/nixdev#188] [Q&A]
I think I might need to wait for either #jkachmar/easy-hls-nix#3 or #127 for this? I made an attempt adding easy-hls-nix to devos like so:

https://github.com/codygman/devos/commit/a15969dd54059a40a7a31f5658463c7b323cede0

I saw evidence it installed things but `haskell-language-server` was not available. I'm just working around it for now by installing via the quickstart nix expression on that readme but it would be nice to know how to pull things in like that repo.

**Comment by Pacman99** [comment]
If its just the package you want to add, you can add it in extern/default.nix. Take a look at how deploy-rs is added, and do something similar. And if you want it in the devshell you can then add it to the packages list within shell/default.nix.

**Comment by nrdxp** [comment]
Additionally, I'm gonna be upgrading the base `nixos` flake to nixos-unstable as 20.09 is getting pretty stale as we approach the next release. However I have to work out an issue with the CI runner hercules-ci/hercules-ci-agent#291 before I can push the update.

Once I do though, hls 1.0 is included in the latest master and already wired up as an override.

**Comment by jkachmar** [comment]
FWIW I'm going to merge jkachmar/easy-hls-nix#4 soon now that it seems like it at least works for the basic TH cases people have pointed out to me.

There's still at least one outstanding issue I've been made aware of with `librt`, but I've only seen someone mention it in the context of GHC 8.6.5 and `reflex-platform` so I may ignore it for the time being.

---

Now that I've read a bit more about `nix run` and Flakes in general I should be able to add the `apps` attribute to my `flake.nix` and support `haskell-language-server-wrapper` that way at the very least (which will close jkachmar/easy-hls-nix#3).

**Reply by blaggacao** [reply]
The `apps` is only necesary if the attribute does not coincide with rhe package name does not coincide with the binary name.

In other cases `nix run .#foo` defaults to `packages.<system>.foo/bin/foo` if no apps declaration is specified.

**Comment by codygman** [comment]
[I can confirm](https://github.com/codygman/devos/commit/8a9b1a6d1917ab9f99615d7f93bf3956603c5b97) that using nixos-unstable and the override that was there in overrides/default.nix haskell-language-server 1.0 is installed :smile: 

I don't understand why just adding `haskell-language-server` installs haskell-language-server 0.9 though.

Wait... is the override silently failing maybe? Since haskell-language-server is a part of each version of ghc's `haskellPackages`? If it had complained loudly I probably could have figured that out, but maybe only having `haskell-language-server` here is somehow valid too... if so I'm not sure what for.

**Reply by codygman** [reply]
Hm, with $work codebase had an issue that either means llvm libs need added to the override or my stack.yaml nix configuration needs to add llvm sources:

```
Installing library in /nix/store/zxn6r5h23niwbkflfiy3pz3phf5262i3-smallcheck-1.1.7/lib/ghc-8.8.4/x86_64-linux-ghc-8.8.4/smallcheck-1.1.7-JtE9utYlx8lL6UNmj2Aao
Setup: '/nix/store/w30rhw13ji6vkv0zksms5h7897cxvl90-ghc-8.8.4/bin/ghc' exited
with an error:
/nix/store/w30rhw13ji6vkv0zksms5h7897cxvl90-ghc-8.8.4/lib/ghc-8.8.4/llvm-targets:
openFile: does not exist (No such file or directory)
```

**Comment by nrdxp** [comment]
At the time of this writing, and since #192 haskell language server should be 1.0 in DevOS even without the override.

**Reply by codygman** [reply]
I haven't fully updated devos and can't till the weekend, I should have prefaced my issue/discussion with that. Sorry! 

My flake inputs are the same though, which is why I was confused about haskell language server being version 0.9 even after `nix flake update --commit-lock-file`. Then again maybe that's PEBKAC too and I need to hurry up and get up to date with divnix/devos to have unbroken `flk update`.

It looks like I'll be able to effortlessly keep up to date with devos with recent changes, thank y'all for your hard work. Hopefully my feedback as someone primarily using devos for day to day work is more useful than nuisance :sweat_smile:

**Reply by nrdxp** [reply]
We absolutely appreciate the feedback, it's a big help. Also, as a heads up, we will be merging #206 fairly soon, hopefully before you update. After than, I'm going to try and get a handle on the filesystem.
